packages/loader/container-loader/src/container.ts

```diff
@@ -2415,83 +2415,86 @@ export class Container
+			if (runtimeFactory === undefined) {
+				throw new Error(packageNotFactoryError);
+			}
 
-		const existing = snapshotTree !== undefined;
+			const existing = snapshotTree !== undefined;
+
+			const context = new ContainerContext(
+				this.options,
+				this.scope,
+				snapshotTree,
+				this._loadedFromVersion,
+				this._deltaManager,
+				this.storageAdapter,
+				this.protocolHandler.quorum,
+				this.protocolHandler.audience,
+				loader,
+				(type, contents, batch, metadata) =>
+					this.submitContainerMessage(type, contents, batch, metadata),
+				(summaryOp: ISummaryContent, referenceSequenceNumber?: number) =>
+					this.submitSummaryMessage(summaryOp, referenceSequenceNumber),
+				(batch: IBatchMessage[], referenceSequenceNumber?: number) =>
+					this.submitBatch(batch, referenceSequenceNumber),
+				(content, targetClientId) => this.submitSignal(content, targetClientId),
+				(error?: ICriticalContainerError) => this.dispose(error),
+				(error?: ICriticalContainerError) => this.close(error),
+				this.updateDirtyContainerState,
+				this.getAbsoluteUrl,
+				() => this.resolvedUrl?.id,
+				() => this.clientId,
+				() => this.attachState,
+				() => this.connected,
+				this._deltaManager.clientDetails,
+				existing,
+				this.subLogger,
+				pendingLocalState,
+				snapshot,
+			);
 
-		const context = new ContainerContext(
-			this.options,
-			this.scope,
-			snapshotTree,
-			this._loadedFromVersion,
-			this._deltaManager,
-			this.storageAdapter,
-			this.protocolHandler.quorum,
-			this.protocolHandler.audience,
-			loader,
-			(type, contents, batch, metadata) =>
-				this.submitContainerMessage(type, contents, batch, metadata),
-			(summaryOp: ISummaryContent, referenceSequenceNumber?: number) =>
-				this.submitSummaryMessage(summaryOp, referenceSequenceNumber),
-			(batch: IBatchMessage[], referenceSequenceNumber?: number) =>
-				this.submitBatch(batch, referenceSequenceNumber),
-			(content, targetClientId) => this.submitSignal(content, targetClientId),
-			(error?: ICriticalContainerError) => this.dispose(error),
-			(error?: ICriticalContainerError) => this.close(error),
-			this.updateDirtyContainerState,
-			this.getAbsoluteUrl,
-			() => this.resolvedUrl?.id,
-			() => this.clientId,
-			() => this.attachState,
-			() => this.connected,
-			this._deltaManager.clientDetails,
-			existing,
-			this.subLogger,
-			pendingLocalState,
-			snapshot,
-		);
+			const runtime = await PerformanceEvent.timedExecAsync(
+				this.subLogger,
+				{ eventName: "InstantiateRuntime" },
+				async () => runtimeFactory.instantiateRuntime(context, existing),
+			);
 
-		const runtime = await PerformanceEvent.timedExecAsync(
-			this.subLogger,
-			{ eventName: "InstantiateRuntime" },
-			async () => runtimeFactory.instantiateRuntime(context, existing),
-		);
+			this._runtime = runtime;
+			this._lifecycleEvents.emit("runtimeInstantiated");
+			this._loadedCodeDetails = codeDetails;
+		} catch (error) {
```

Comment:
@vladsud @anthony-murphy FYI, I added this try catch block around container runtime instantiation because failures during it does not result in container close or dispose. That seems like a bug to me.